From f62e296883689140bef94c80b7cb7fefb8c7dc23 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Wed, 9 Nov 2016 14:30:09 -0800
Subject: [PATCH] MORAY-334 minnow clients reporting no active connections when
 moray seems to be up MANTA-2972 minnow needs to lose no-op moray error
 handlers MANTA-3020 minnow should relax about transient failures MANTA-3021
 minnow should abort on uncaught exceptions MANTA-3022 minnow repo needs a
 little cleanup Reviewed by: Robert Mustacchi <rm@joyent.com> Approved by:
 Robert Mustacchi <rm@joyent.com>

---
 Makefile                    |  4 ++--
 lib/index.js                | 15 ---------------
 main.js                     | 33 +++++++++++----------------------
 package.json                |  2 +-
 smf/manifests/minnow.xml.in |  2 +-
 5 files changed, 15 insertions(+), 41 deletions(-)
 delete mode 100644 lib/index.js

diff --git a/Makefile b/Makefile
index fcaa448..a997378 100644
--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,7 @@
 # Files
 #
 DOC_FILES	 = index.md
-JS_FILES	:= $(shell ls *.js) $(shell find lib -name '*.js')
+JS_FILES	:= $(shell ls *.js)
 JSL_CONF_NODE	 = tools/jsl.node.conf
 JSL_FILES_NODE   = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
@@ -39,7 +39,7 @@ SMF_MANIFESTS_IN = smf/manifests/minnow.xml.in
 
 NAME 			= minnow
 NODE_PREBUILT_TAG	= zone
-NODE_PREBUILT_VERSION	:= v0.10.25
+NODE_PREBUILT_VERSION	:= v0.10.48
 NODE_PREBUILT_IMAGE	= fd2cc906-8938-11e3-beab-4359c665ac99
 
 include ./tools/mk/Makefile.defs
diff --git a/lib/index.js b/lib/index.js
deleted file mode 100644
index 81111d7..0000000
--- a/lib/index.js
+++ /dev/null
@@ -1,15 +0,0 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2014, Joyent, Inc.
- */
-
-/*
- * A brief overview of this source file: what is its purpose.
- */
-
-// TODO: put stuff here
diff --git a/main.js b/main.js
index e101bad..f77b381 100644
--- a/main.js
+++ b/main.js
@@ -53,7 +53,8 @@ var HEARTBEAT;
 var INTERVAL;
 var TIMER;
 var LAST_ERROR;
-
+var NERRORS = 0;
+var NERROR_THRESHOLD = 3;
 
 
 ///--- CLI Functions
@@ -137,7 +138,6 @@ function createMorayClient(opts, cb) {
     cb = once(cb);
 
     var retry = opts.retry || {};
-    retry.retries = retry.retries || Infinity;
     retry.minTimeout = retry.minTimeout || 2000;
     retry.maxTimeout = retry.maxTimeout || 120000;
     var client = moray.createClient({
@@ -149,40 +149,24 @@ function createMorayClient(opts, cb) {
         retry: retry
     });
 
-    function onConnectError(err) {
-        LOG.error(err, 'moray: connect: failed; will retry in 5 seconds');
-        client.removeListener('connect', onConnectSetup);
-        client.close();
-        setTimeout(createMorayClient.bind(null, opts, cb), 5000);
-    }
-
     function onConnectSetup() {
         var bname = opts.bucket.name;
         var index = opts.bucket.index;
 
-        client.removeListener('error', onConnectError);
-
-        client.on('error', function (err) {
-            // not much more to do because the moray client should take
-            // care of reconnecting, etc.
-            LOG.error(err, 'moray client error');
-        });
-
+        LOG.info('moray: connected, setting up buckets');
         client.putBucket(bname, {index: index}, function (err) {
             if (err) {
                 LOG.error(err, 'moray: putBucket: failed; will retry in 5 ' +
                           'seconds');
-                client.close();
-                setTimeout(createMorayClient.bind(null, opts, cb), 5000);
+                setTimeout(onConnectSetup, 5000);
             } else {
-                LOG.info('moray: connected');
+                LOG.info('moray: buckets set up');
                 cb(null, client);
             }
         });
     }
 
     client.once('connect', onConnectSetup);
-    client.once('error', onConnectError);
 }
 
 
@@ -236,11 +220,16 @@ function heartbeat(opts) {
 
         opts.moray.putObject(opts.bucket, key, stats, function (err) {
             if (err) {
-                LOG.error(err, 'moray: update failed');
+                if (++NERRORS >= NERROR_THRESHOLD) {
+                    LOG.error(err, 'moray: update failed');
+                } else {
+                    LOG.warn(err, 'moray: update failed');
+                }
                 LAST_ERROR = err;
                 return;
             }
 
+            NERRORS = 0;
             LAST_ERROR = null;
             LOG.info({
                 bucket: opts.bucket,
diff --git a/package.json b/package.json
index 5cd6b5a..d934fc4 100644
--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
         "bunyan": "0.21.4",
         "dashdash": "1.3.2",
         "libmanta": "git+ssh://git@github.com:joyent/node-libmanta.git#master",
-        "moray": "git+ssh://git@github.com:joyent/node-moray.git#fd5781bc25a9bfe2ba82167664639753fb9f0ca5",
+        "moray": "^2.0.0",
         "node-uuid": "1.4.0",
         "once": "1.2.0",
         "statvfs": "2.0.0"
diff --git a/smf/manifests/minnow.xml.in b/smf/manifests/minnow.xml.in
index 641fdcf..388d227 100644
--- a/smf/manifests/minnow.xml.in
+++ b/smf/manifests/minnow.xml.in
@@ -46,7 +46,7 @@
 
 	<exec_method type="method"
 		     name="start"
-		     exec="node main.js -f ./etc/config.json &amp;"
+		     exec="node --abort-on-uncaught-exception main.js -f ./etc/config.json &amp;"
 		     timeout_seconds="30">
 	    <method_context working_directory="/opt/smartdc/minnow">
                 <method_credential user="nobody"
-- 
2.21.0

